logical/src/main/java/org/apache/drill/common/expression/fn/JodaDateValidator.java

```diff
@@ -0,0 +1,216 @@
+  public static final String JODA_FULL_NAME_OF_DAY = "EEEE";
+  public static final String JODA_DAY_OF_YEAR = "D";
+  public static final String JODA_DAY_OF_MONTH = "d";
+  public static final String JODA_DAY_OF_WEEK = "e";
+  public static final String JODA_NAME_OF_MONTH = "MMMM";
+  public static final String JODA_ABR_NAME_OF_MONTH = "MMM";
+  public static final String JODA_FULL_ERA_NAME = "G";
+  public static final String JODA_NAME_OF_DAY = "E";
+  public static final String JODA_TIME_ZONE_NAME = "TZ";
+  public static final String JODA_HOUR_12_NAME = "h";
+  public static final String JODA_HOUR_12_OTHER_NAME = "h";
+  public static final String JODA_HOUR_24_NAME = "H";
+  public static final String JODA_MINUTE_OF_HOUR_NAME = "m";
+  public static final String JODA_SECOND_OF_MINUTE_NAME = "s";
+  public static final String JODA_MILLISECOND_OF_MINUTE_NAME = "SSS";
+  public static final String JODA_WEEK_OF_YEAR = "w";
+  public static final String JODA_MONTH = "MM";
+  public static final String JODA_HALFDAY = "aa";
+
+  static {
+    postgresToJodaMap.put(POSTGRES_FULL_NAME_OF_DAY, JODA_FULL_NAME_OF_DAY);
+    postgresToJodaMap.put(POSTGRES_DAY_OF_YEAR, JODA_DAY_OF_YEAR);
+    postgresToJodaMap.put(POSTGRES_DAY_OF_MONTH, JODA_DAY_OF_MONTH);
+    postgresToJodaMap.put(POSTGRES_DAY_OF_WEEK, JODA_DAY_OF_WEEK);
+    postgresToJodaMap.put(POSTGRES_NAME_OF_MONTH, JODA_NAME_OF_MONTH);
+    postgresToJodaMap.put(POSTGRES_ABR_NAME_OF_MONTH, JODA_ABR_NAME_OF_MONTH);
+    postgresToJodaMap.put(POSTGRES_FULL_ERA_NAME, JODA_FULL_ERA_NAME);
+    postgresToJodaMap.put(POSTGRES_NAME_OF_DAY, JODA_NAME_OF_DAY);
+    postgresToJodaMap.put(POSTGRES_TIME_ZONE_NAME, JODA_TIME_ZONE_NAME);
+    postgresToJodaMap.put(POSTGRES_HOUR_12_NAME, JODA_HOUR_12_NAME);
+    postgresToJodaMap.put(POSTGRES_HOUR_12_OTHER_NAME, JODA_HOUR_12_OTHER_NAME);
+    postgresToJodaMap.put(POSTGRES_HOUR_24_NAME, JODA_HOUR_24_NAME);
+    postgresToJodaMap.put(POSTGRES_MINUTE_OF_HOUR_NAME, JODA_MINUTE_OF_HOUR_NAME);
+    postgresToJodaMap.put(POSTGRES_SECOND_OF_MINUTE_NAME, JODA_SECOND_OF_MINUTE_NAME);
+    postgresToJodaMap.put(POSTGRES_MILLISECOND_OF_MINUTE_NAME, JODA_MILLISECOND_OF_MINUTE_NAME);
+    postgresToJodaMap.put(POSTGRES_WEEK_OF_YEAR, JODA_WEEK_OF_YEAR);
+    postgresToJodaMap.put(POSTGRES_MONTH, JODA_MONTH);
+    postgresToJodaMap.put(POSTGRES_HALFDAY_AM, JODA_HALFDAY);
+    postgresToJodaMap.put(POSTGRES_HALFDAY_PM, JODA_HALFDAY);
+  }
+
+  static {
+    postgresValuesForDeleting.add(SUFFIX_SP);
+    postgresValuesForDeleting.add(PREFIX_FM);
+    postgresValuesForDeleting.add(PREFIX_FX);
+    postgresValuesForDeleting.add(PREFIX_TM);
+  }
+
+  /**
+   * Replaces all postgres patterns from {@param pattern},
+   * available in postgresToJodaMap keys to jodaTime equivalents.
+   *
+   * @param pattern date pattern in postgres format
+   * @return date pattern with replaced patterns in joda format
+   */
+  public static String toJodaFormat(String pattern) {
```

Comment:
The premise of this function is that no one pattern is a subset of another pattern. And, no one pattern is the same as the same as part of two adjacent patterns. That is, we know that, say, SSSS is a single pattern and not two copies of SS next to one another. Have we done the analysis to ensure that no such aliases are possible?

Or, for example, consider these patterns: DAY, Y, YY, YYY, IYY and IYYY. If we replace "Y" first, we end up matching "DAY" or "IYY" when we should not. If we do "IY" first, we miss "IYYY". If the input is "DAYY", and we do "YY" first, we miss that this is "DAY - Y", not "DA - YY". In short, context seems essential. Or, the substitutions must be carefully ordered. I think that, for the above, "IYYY", "IYY", "YYY", "YY", "DAY", "Y" would be the proper order. But, since we use a map, do we have an implied replacement order?